MuxExoPlayer/src/main/java/com/mux/stats/sdk/muxstats/MuxBaseExoPlayer.java

```diff
@@ -779,251 +788,199 @@ public void outputLog(String tag, String msg) {
         return;
       }
-      BandwidthMetricData loadData = currentBandwidthMetric().onLoadCanceled(dataSpec);
-      dispatch(loadData);
+      BandwidthMetricData loadData = currentBandwidthMetric().onLoadCanceled(segmentUrl);
+      parseHeaders(loadData, headers);
+      dispatch(loadData, new RequestCanceled(null));
     }
 
-    public void onLoadStarted(DataSpec dataSpec, int dataType, Format trackFormat,
-        long mediaStartTimeMs, long mediaEndTimeMs, long elapsedRealtimeMs) {
+    public void onLoadStarted(long mediaStartTimeMs, long mediaEndTimeMs, String segmentUrl,
+        int dataType, String host) {
       if (player == null || player.get() == null || muxStats == null
           || currentBandwidthMetric() == null) {
         return;
       }
-      currentBandwidthMetric().onLoadStarted(dataSpec, dataType,
-          trackFormat, mediaStartTimeMs,
-          mediaEndTimeMs, elapsedRealtimeMs);
+      currentBandwidthMetric().onLoadStarted(mediaStartTimeMs, mediaEndTimeMs, segmentUrl
+          , dataType, host);
     }
 
-    public void onLoadCompleted(DataSpec dataSpec, int dataType, Format trackFormat,
-        long mediaStartTimeMs, long mediaEndTimeMs, long elapsedRealtimeMs, long loadDurationMs,
-        long bytesLoaded, Map<String, List<String>> responseHeaders) {
+    public void onLoadCompleted(
+        String segmentUrl, long bytesLoaded, Format trackFormat,
+        Map<String, List<String>> responseHeaders) {
       if (player == null || player.get() == null || muxStats == null
           || currentBandwidthMetric() == null) {
         return;
       }
-      BandwidthMetricData loadData = currentBandwidthMetric().onLoadCompleted(dataSpec,
-          dataType,
-          trackFormat, mediaStartTimeMs,
-          mediaEndTimeMs, elapsedRealtimeMs, loadDurationMs, bytesLoaded);
+      BandwidthMetricData loadData = currentBandwidthMetric().onLoadCompleted(
+          segmentUrl, bytesLoaded, trackFormat);
+      parseHeaders(loadData, responseHeaders);
+      dispatch(loadData, new RequestCompleted(null));
+    }
 
-      // Only append this data if we have some load data going on already
-      // TODO - this does not work correctly today, fix this and re-enable it.
-//            if (loadData != null && responseHeaders != null) {
-//                Hashtable<String, String> headers = parseHeaders(responseHeaders);
-//
-//                if (headers != null) {
-//                    loadData.setRequestResponseHeaders(headers);
-//                }
-//            }
+    private void parseHeaders(BandwidthMetricData loadData,
+        Map<String, List<String>> responseHeaders) {
+      if (responseHeaders != null) {
+        Hashtable<String, String> headers = parseHeaders(responseHeaders);
 
-      dispatch(loadData);
+        if (headers != null) {
+          loadData.setRequestResponseHeaders(headers);
+        }
+      } else {
+        MuxLogger.d(TAG, "ERROR: No headers found for segment !!!");
```

Comment:
I don't think we need to log this; sometimes there won't be headers for us to parse, and it's fine to just silently ignore this because it will happen a lot in normal use.

Reply:
fixed